### ISSUE TITLE
 java.lang.StackOverflowError with retrofit 2 @URL based

### ISSUE THREAD

**Issue by TheGreat0004** [square/retrofit#1959]
FATAL EXCEPTION: OkHttp Dispatcher
                                                           java.lang.StackOverflowError
                                                               at org.apache.harmony.security.asn1.ASN1StringType.getDecodedObject(ASN1StringType.java:99)
                                                               at org.apache.harmony.security.asn1.ASN1StringType.decode(ASN1StringType.java:90)
                                                               at org.apache.harmony.security.asn1.ASN1Implicit.decode(ASN1Implicit.java:116)
                                                               at org.apache.harmony.security.asn1.ASN1Choice.decode(ASN1Choice.java:308)
                                                               at org.apache.harmony.security.asn1.BerInputStream.decodeValueCollection(BerInputStream.java:626)
                                                               at org.apache.harmony.security.asn1.BerInputStream.readSequenceOf(BerInputStream.java:584)
                                                               at org.apache.harmony.security.asn1.ASN1SequenceOf.decode(ASN1SequenceOf.java:40)
                                                               at org.apache.harmony.security.asn1.ASN1Type.decode(ASN1Type.java:82)
                                                               at org.apache.harmony.security.x509.Extensions.decodeGeneralNames(Extensions.java:292)
                                                               at org.apache.harmony.security.x509.Extensions.valueOfSubjectAlternativeName(Extensions.java:260)
                                                               at org.apache.harmony.security.provider.cert.X509CertImpl.getSubjectAlternativeNames(X509CertImpl.java:321)
                                                               at okhttp3.internal.tls.OkHostnameVerifier.getSubjectAltNames(OkHostnameVerifier.java:109)
                                                               at okhttp3.internal.tls.OkHostnameVerifier.verifyHostname(OkHostnameVerifier.java:77)
                                                               at okhttp3.internal.tls.OkHostnameVerifier.verify(OkHostnameVerifier.java:59)
                                                               at okhttp3.internal.tls.OkHostnameVerifier.verify(OkHostnameVerifier.java:50)
                                                               at okhttp3.internal.io.RealConnection.connectTls(RealConnection.java:199)
                                                               at okhttp3.internal.io.RealConnection.connectSocket(RealConnection.java:148)
                                                               at okhttp3.internal.io.RealConnection.connect(RealConnection.java:111)
                                                               at okhttp3.internal.http.StreamAllocation.findConnection(StreamAllocation.java:188)
                                                               at okhttp3.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:127)
                                                               at okhttp3.internal.http.StreamAllocation.newStream(StreamAllocation.java:97)
                                                               at okhttp3.internal.http.HttpEngine.connect(HttpEngine.java:289)
                                                               at okhttp3.internal.http.HttpEngine.sendRequest(HttpEngine.java:241)
                                                               at okhttp3.RealCall.getResponse(RealCall.java:240)
                                                               at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:198)
                                                               at cl.dropcp.controllers.rest.RetrofitAdapters$1.intercept(RetrofitAdapters.java:57)
                                                               at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:187)
                                                               at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.java:203)
                                                               at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:187)
                                                               at cl.dropcp.controllers.rest.RetrofitAdapters$1.intercept(RetrofitAdapters.java:57)
                                                               at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:187)
                                                               at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.java:203)
                                                               at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:187)
                                                               at cl.dropcp.controllers.rest.RetrofitAdapters$1.intercept(RetrofitAdapters.java:57)
                                                               at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:187)
                                                               at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.java:203)
                                                               at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:187)
                                                               at cl.dropcp.controllers.rest.RetrofitAdapters$1.intercept(RetrofitAdapters.java:57)
                                                               at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:187)
                                                               at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.java:203)
                                                               at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:187)
                                                               at cl.dropcp.controllers.rest.RetrofitAdapters$1.intercept(RetrofitAdapters.java:57)
                                                               at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:187)
                                                               at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.java:203)
                                                               at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:187)
                                                               at cl.dropcp.controllers.rest.RetrofitAdapters$1.intercept(RetrofitAdapters.java:57)
                                                               at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:187)
                                                               at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.java:203)
                                                           	at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:1

This is my adapter,
public class RetrofitAdapters {
    private static  OkHttpClient.Builder CLIENT;
    private static String ROOT =BuildConfig.HOST + "/api/driver/";

    static {
        CLIENT = GeneralFunction.getOkHttpClient();
    }

    @NonNull
    public static <S> S createRetrofitService(Class<S> serviceClass) {
        return getRetrofit().create(serviceClass);
    }

    @NonNull
    private static Retrofit getRetrofit() {

        if (BuildConfig.DEBUG) {
            HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
            interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
            CLIENT.addInterceptor(interceptor);
        }

        if (CLIENT.interceptors().isEmpty())
            CLIENT.addInterceptor(new Interceptor() {
                @Override
                public okhttp3.Response intercept(Chain chain) throws IOException {

                    Request request = chain.request();
                    String token = ApplicationClass.getAccessToken();
                    if (token != null && !token.isEmpty())
                        request = request.newBuilder().addHeader(RestConstants.AUTHORIZATION, "Bearer "+token).build();
                    return chain.proceed(request);
                }
            });
        return new Retrofit.Builder()
                .baseUrl(ROOT)
                .addConverterFactory(JacksonConverterFactory.create())
                .client(CLIENT.build())
                .build();
    }

    public static MErrorHandler parseError(Response<?> response) {
        Converter<ResponseBody, MErrorHandler> converter =
                getRetrofit().responseBodyConverter(MErrorHandler.class, new Annotation[0]);

        MErrorHandler error;

        try {
            error = converter.convert(response.errorBody());
        } catch (IOException e) {
            return new MErrorHandler();
        }

        return error;
    }
}

Basically i have added interceptor as with every request, requires an accessToken but in some independent request it didn't require it, and my app randomly crash there.

**Comment by JakeWharton** [comment]
Every time you call `getRetrofit()` you're adding interceptors to the `OkHttpClient.Builder` so eventually you'll add so many that you overflow the stack. Create the `OkHttpClient.Builder` once, call `.build()`, and create the `Retrofit` instance _once_, and then re-use that instance over and over instead of creating new ones each time.

**Comment by TheGreat0004** [comment]
@JakeWharton , how can i do that as the token can be null, due to which i am adding tit multiple times....
